
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>utils &#8212; pdb_prot_align 0.3.0 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Package index" href="package_index.html" />
    <link rel="prev" title="pymol_struct" href="pdb_prot_align.pymol_struct.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <span class="target" id="module-pdb_prot_align.utils"></span><div class="section" id="utils">
<h1>utils<a class="headerlink" href="#utils" title="Permalink to this headline">¶</a></h1>
<p>Utilities for aligning proteins to PDB.</p>
<dl class="function">
<dt id="pdb_prot_align.utils.align_prots_mafft">
<code class="sig-prename descclassname">pdb_prot_align.utils.</code><code class="sig-name descname">align_prots_mafft</code><span class="sig-paren">(</span><em class="sig-param">prots</em>, <em class="sig-param">*</em>, <em class="sig-param">mafft='mafft'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pdb_prot_align/utils.html#align_prots_mafft"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pdb_prot_align.utils.align_prots_mafft" title="Permalink to this definition">¶</a></dt>
<dd><p>Align protein sequences with <code class="docutils literal notranslate"><span class="pre">mafft</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>prots</strong> (<em>list</em>) – List of sequences as <cite>Biopython.SeqRecord.SeqRecord</cite> objects.</p></li>
<li><p><strong>alignmentfile</strong> (<em>str</em>) – Name of created FASTA alignment.</p></li>
<li><p><strong>mafft</strong> (<em>str</em>) – Command that resolves to <code class="docutils literal notranslate"><span class="pre">mafft</span></code> executable.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The protein alignment.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Bio.Align.MultipleSeqAlignment</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pdb_prot_align.utils.aligned_site_map">
<code class="sig-prename descclassname">pdb_prot_align.utils.</code><code class="sig-name descname">aligned_site_map</code><span class="sig-paren">(</span><em class="sig-param">aln</em>, <em class="sig-param">heads</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pdb_prot_align/utils.html#aligned_site_map"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pdb_prot_align.utils.aligned_site_map" title="Permalink to this definition">¶</a></dt>
<dd><p>Get mapping between sequential site numbers of aligned sequences.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>aln</strong> (<em>Bio.Align.MultipleSeqAlignment</em>) – Alignment of sequences</p></li>
<li><p><strong>heads</strong> (<em>list</em>) – Headers of sequences of interest in <cite>aln</cite>, must be at least two.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Columns are sequence names in <cite>heads</cite>. There is a row for alignment
positions that are non-gapped in at least one sequence in <cite>heads</cite>, and
the entry in the row gives the number of that site in 1, 2, …
numbering of the sequence, or <cite>NaN</cite> if there is no residue for
sequence at that position in alignment.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
<p class="rubric">Example</p>
<p>Create and parse an alignment:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">TemporaryFile</span><span class="p">(</span><span class="s1">&#39;w+&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">textwrap</span><span class="o">.</span><span class="n">dedent</span><span class="p">(</span>
<span class="gp">... </span>            <span class="sd">&#39;&#39;&#39;</span>
<span class="gp">... </span><span class="sd">            &gt;seq1</span>
<span class="gp">... </span><span class="sd">            GK-AC-L</span>
<span class="gp">... </span><span class="sd">            &gt;seq2</span>
<span class="gp">... </span><span class="sd">            -K-AM-L</span>
<span class="gp">... </span><span class="sd">            &gt;seq3</span>
<span class="gp">... </span><span class="sd">            G-NAC-L</span>
<span class="gp">... </span><span class="sd">            &gt;seq4</span>
<span class="gp">... </span><span class="sd">            GK-ACSI</span>
<span class="gp">... </span><span class="sd">            &#39;&#39;&#39;</span>
<span class="gp">... </span>            <span class="p">))</span>
<span class="gp">... </span>    <span class="n">f</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">aln</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Show the alignment site map. The gymnastics with converting to str and
replacing ‘nan’ are to make output format the same in both <cite>pandas</cite>
&gt;= and &lt; 1.0 and earlier versions so doctest passes in both:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">aligned_site_map</span><span class="p">(</span><span class="n">aln</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;seq1&#39;</span><span class="p">,</span> <span class="s1">&#39;seq2&#39;</span><span class="p">,</span> <span class="s1">&#39;seq3&#39;</span><span class="p">])</span>
<span class="gp">... </span> <span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;nan&#39;</span><span class="p">,</span> <span class="s1">&#39;&lt;NA&gt;&#39;</span><span class="p">))</span>
<span class="go">   seq1  seq2  seq3</span>
<span class="go">0     1  &lt;NA&gt;     1</span>
<span class="go">1     2     1  &lt;NA&gt;</span>
<span class="go">2  &lt;NA&gt;  &lt;NA&gt;     2</span>
<span class="go">3     3     2     3</span>
<span class="go">4     4     3     4</span>
<span class="go">5     5     4     5</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pdb_prot_align.utils.alignment_to_count_df">
<code class="sig-prename descclassname">pdb_prot_align.utils.</code><code class="sig-name descname">alignment_to_count_df</code><span class="sig-paren">(</span><em class="sig-param">fasta_alignment</em>, <em class="sig-param">*</em>, <em class="sig-param">ignore_gaps=False</em>, <em class="sig-param">as_freqs=False</em>, <em class="sig-param">add_entropy_neff=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pdb_prot_align/utils.html#alignment_to_count_df"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pdb_prot_align.utils.alignment_to_count_df" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert FASTA alignment to counts data frame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fasta_alignment</strong> (<em>str</em>) – FASTA file with alignment.</p></li>
<li><p><strong>ignore_gaps</strong> (<em>bool</em>) – Ignore gap characters (‘-‘) in counts.</p></li>
<li><p><strong>as_freqs</strong> (<em>bool</em>) – Return frequencies rather than counts.</p></li>
<li><p><strong>add_entropy_neff</strong> (<em>bool</em>) – Add columns with site entropy (bits) and number effective amino acids.
Requires <cite>as_freqs</cite> to be <cite>True</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>First column (‘isite’) is 1, 2, … site numbering. Other columns give
counts of each observed alignment character at that site.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">NamedTemporaryFile</span><span class="p">(</span><span class="s1">&#39;w+&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">textwrap</span><span class="o">.</span><span class="n">dedent</span><span class="p">(</span>
<span class="gp">... </span>            <span class="sd">&#39;&#39;&#39;</span>
<span class="gp">... </span><span class="sd">            &gt;seq1</span>
<span class="gp">... </span><span class="sd">            GK-AC-L</span>
<span class="gp">... </span><span class="sd">            &gt;seq2</span>
<span class="gp">... </span><span class="sd">            -K-AM-L</span>
<span class="gp">... </span><span class="sd">            &gt;seq3</span>
<span class="gp">... </span><span class="sd">            G-NAC-L</span>
<span class="gp">... </span><span class="sd">            &gt;seq4</span>
<span class="gp">... </span><span class="sd">            GK-ACSI</span>
<span class="gp">... </span><span class="sd">            &#39;&#39;&#39;</span>
<span class="gp">... </span>            <span class="p">))</span>
<span class="gp">... </span>    <span class="n">f</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">counts</span> <span class="o">=</span> <span class="n">alignment_to_count_df</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">counts_nogaps</span> <span class="o">=</span> <span class="n">alignment_to_count_df</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">ignore_gaps</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">freqs_nogaps</span> <span class="o">=</span> <span class="n">alignment_to_count_df</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
<span class="gp">... </span>                                         <span class="n">ignore_gaps</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span>                                         <span class="n">as_freqs</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">freqs_with_ent</span> <span class="o">=</span> <span class="n">alignment_to_count_df</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
<span class="gp">... </span>                                           <span class="n">ignore_gaps</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span>                                           <span class="n">as_freqs</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span>                                           <span class="n">add_entropy_neff</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="gp">... </span>                                           <span class="p">)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span>
<span class="go">   isite  -  A  C  G  I  K  L  M  N  S</span>
<span class="go">0      1  1  0  0  3  0  0  0  0  0  0</span>
<span class="go">1      2  1  0  0  0  0  3  0  0  0  0</span>
<span class="go">2      3  3  0  0  0  0  0  0  0  1  0</span>
<span class="go">3      4  0  4  0  0  0  0  0  0  0  0</span>
<span class="go">4      5  0  0  3  0  0  0  0  1  0  0</span>
<span class="go">5      6  3  0  0  0  0  0  0  0  0  1</span>
<span class="go">6      7  0  0  0  0  1  0  3  0  0  0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts_nogaps</span>
<span class="go">   isite  A  C  G  I  K  L  M  N  S</span>
<span class="go">0      1  0  0  3  0  0  0  0  0  0</span>
<span class="go">1      2  0  0  0  0  3  0  0  0  0</span>
<span class="go">2      3  0  0  0  0  0  0  0  1  0</span>
<span class="go">3      4  4  0  0  0  0  0  0  0  0</span>
<span class="go">4      5  0  3  0  0  0  0  1  0  0</span>
<span class="go">5      6  0  0  0  0  0  0  0  0  1</span>
<span class="go">6      7  0  0  0  1  0  3  0  0  0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">freqs_nogaps</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="go">   isite    A     C    G     I    K     L     M    N    S</span>
<span class="go">0      1  0.0  0.00  1.0  0.00  0.0  0.00  0.00  0.0  0.0</span>
<span class="go">1      2  0.0  0.00  0.0  0.00  1.0  0.00  0.00  0.0  0.0</span>
<span class="go">2      3  0.0  0.00  0.0  0.00  0.0  0.00  0.00  1.0  0.0</span>
<span class="go">3      4  1.0  0.00  0.0  0.00  0.0  0.00  0.00  0.0  0.0</span>
<span class="go">4      5  0.0  0.75  0.0  0.00  0.0  0.00  0.25  0.0  0.0</span>
<span class="go">5      6  0.0  0.00  0.0  0.00  0.0  0.00  0.00  0.0  1.0</span>
<span class="go">6      7  0.0  0.00  0.0  0.25  0.0  0.75  0.00  0.0  0.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">freqs_with_ent</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;entropy&#39;</span><span class="p">,</span> <span class="s1">&#39;n_effective&#39;</span><span class="p">])</span> <span class="o">==</span>
<span class="gp">... </span> <span class="n">freqs_nogaps</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">freqs_with_ent</span><span class="p">[[</span><span class="s1">&#39;entropy&#39;</span><span class="p">,</span> <span class="s1">&#39;n_effective&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="go">   entropy  n_effective</span>
<span class="go">0    0.000        1.000</span>
<span class="go">1    0.000        1.000</span>
<span class="go">2    0.000        1.000</span>
<span class="go">3    0.000        1.000</span>
<span class="go">4    0.811        1.755</span>
<span class="go">5    0.000        1.000</span>
<span class="go">6    0.811        1.755</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="pdb_prot_align.utils.pdb_seq_to_number">
<code class="sig-prename descclassname">pdb_prot_align.utils.</code><code class="sig-name descname">pdb_seq_to_number</code><span class="sig-paren">(</span><em class="sig-param">pdbfile</em>, <em class="sig-param">chain_ids</em>, <em class="sig-param">chain_identity='require_same'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pdb_prot_align/utils.html#pdb_seq_to_number"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pdb_prot_align.utils.pdb_seq_to_number" title="Permalink to this definition">¶</a></dt>
<dd><p>Get sequence and number of chain(s) for some protein in PDB file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pdbfile</strong> (<em>str</em>) – Path to existing PDB file.</p></li>
<li><p><strong>chain_ids</strong> (<em>list</em>) – List of chains in PDB file. All these chains must correspond to the
same underlying molecule (i.e., be monomers in homo-oligomer).</p></li>
<li><p><strong>chain_identity</strong> (<em>{'union'</em><em>, </em><em>'intersection'</em><em>, </em><em>'require_same'}</em>) – How to parse chains. They are required to share the same wildtype
at all sites they have in common. If all sites are not shared
between all chains, take the union, the intersection, or raise
an error if they are not exactly the same.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="simple">
<dt>Columns are:</dt><dd><ul class="simple">
<li><p>’sequential’ : sequential 1, 2, … numbering of residues</p></li>
<li><p>’pdb_site’ : PDB file residue number</p></li>
<li><p>’wildtype’ : wildtype residue identity (1-letter code)</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="pdb_prot_align.utils.strip_gaps_to_ref">
<code class="sig-prename descclassname">pdb_prot_align.utils.</code><code class="sig-name descname">strip_gaps_to_ref</code><span class="sig-paren">(</span><em class="sig-param">aln</em>, <em class="sig-param">ref_head</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pdb_prot_align/utils.html#strip_gaps_to_ref"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pdb_prot_align.utils.strip_gaps_to_ref" title="Permalink to this definition">¶</a></dt>
<dd><p>Strip gaps in alignment relative to reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>aln</strong> (<em>Bio.Align.MultipleSeqAlignment</em>) – The alignment of sequences.</p></li>
<li><p><strong>ref_head</strong> (<em>str</em>) – Header of reference sequence in <cite>aln</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Alignment as list of  2-tuples of strings (<cite>header, sequence</cite>) with
all gaps relative to reference removed.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
<p class="rubric">Example</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">with</span> <span class="n">tempfile</span><span class="o">.</span><span class="n">TemporaryFile</span><span class="p">(</span><span class="s1">&#39;w+&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">textwrap</span><span class="o">.</span><span class="n">dedent</span><span class="p">(</span>
<span class="gp">... </span>            <span class="sd">&#39;&#39;&#39;</span>
<span class="gp">... </span><span class="sd">            &gt;seq1</span>
<span class="gp">... </span><span class="sd">            GK-AC-L</span>
<span class="gp">... </span><span class="sd">            &gt;seq2</span>
<span class="gp">... </span><span class="sd">            -K-AM-L</span>
<span class="gp">... </span><span class="sd">            &gt;seq3</span>
<span class="gp">... </span><span class="sd">            G-NAC-L</span>
<span class="gp">... </span><span class="sd">            &gt;seq4</span>
<span class="gp">... </span><span class="sd">            GK-ACSI</span>
<span class="gp">... </span><span class="sd">            &#39;&#39;&#39;</span>
<span class="gp">... </span>            <span class="p">))</span>
<span class="gp">... </span>    <span class="n">f</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
<span class="gp">... </span>    <span class="n">_</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">... </span>    <span class="n">aln</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">strip_gaps_to_ref</span><span class="p">(</span><span class="n">aln</span><span class="p">,</span> <span class="s1">&#39;seq1&#39;</span><span class="p">)</span>  
<span class="go">[(&#39;seq1&#39;, &#39;GKACL&#39;),</span>
<span class="go"> (&#39;seq2&#39;, &#39;-KAML&#39;),</span>
<span class="go"> (&#39;seq3&#39;, &#39;G-ACL&#39;),</span>
<span class="go"> (&#39;seq4&#39;, &#39;GKACI&#39;)]</span>
</pre></div>
</div>
</dd></dl>

</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/BloomLogo.jpg" alt="Logo"/>
    
    <h1 class="logo logo-name">pdb_prot_align</h1>
    
  </a>
</p>



<p class="blurb">Align proteins to PDB chains</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=jbloomlab&repo=pdb_prot_align&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





    

<p>
<a class="badge" href="https://travis-ci.org/jbloomlab/pdb_prot_align">
    <img
        alt="https://secure.travis-ci.org/jbloomlab/pdb_prot_align.svg?branch=master"
        src="https://secure.travis-ci.org/jbloomlab/pdb_prot_align.svg?branch=master"
    />
</a>
</p>


<h3>Navigation</h3>
<ul>
<li class="toctree-l1"><a class="reference internal" href="index.html">pdb_prot_align documentation</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">Command-line interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="pdb_prot_align.html">pdb_prot_align package</a></li>
<li class="toctree-l1"><a class="reference internal" href="package_index.html">Package index</a></li>
<li class="toctree-l1"><a class="reference internal" href="acknowledgments.html">Acknowledgements</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="pdb_prot_align.html">pdb_prot_align package</a><ul>
      <li>Previous: <a href="pdb_prot_align.pymol_struct.html" title="previous chapter">pymol_struct</a></li>
      <li>Next: <a href="package_index.html" title="next chapter">Package index</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019--2020.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.4.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/pdb_prot_align.utils.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/jbloomlab/pdb_prot_align" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>